### PR TITLE
add gVNIC support for compute instance

### PR DESCRIPTION
**Pull request by edwardmedia**
Fixes https://github.com/hashicorp/terraform-provider-google/issues/7699

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [ ] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
compute: added `nic_type` field to `google_compute_instance` resource to support gVNIC
```

```release-note:enhancement
compute: added `nic_type` field to `google_compute_instance_template ` resource to support gVNIC
```

### CODE REVIEW

**Review by slevenick — CHANGES_REQUESTED** [review]
Seeing lots of failures on: 

```
panic: interface conversion: interface {} is nil, not string
goroutine 258 [running]:
github.com/hashicorp/terraform-provider-google-beta/v3/google-beta.expandNetworkInterfaces(0x3bbcb40, 0xc001049500, 0xc000753800, 0x0, 0x0, 0x0, 0x0, 0x0)
  /terraform-provider-google-beta/google-beta/compute_instance_helpers.go:253 +0xc82
```

File: mmv1/third_party/terraform/utils/compute_instance_helpers.go.erb [hunk 1]
```diff
@@ -254,6 +255,7 @@ func expandNetworkInterfaces(d TerraformResourceData, config *Config) ([]*comput
 			Subnetwork:    sf.RelativeLink(),
 			AccessConfigs: expandAccessConfigs(data["access_config"].([]interface{})),
 			AliasIpRanges: expandAliasIpRanges(data["alias_ip_range"].([]interface{})),
+			NicType:       data["nic_type"].(string),
```
Comment (slevenick) [inline mmv1/third_party/terraform/utils/compute_instance_helpers.go.erb:258]: Looks like we need different handling of this field when nic_type is not set. I believe it is coming back as nil which cannot be type asserted to a string. You may want to try something like:

Probably want to check for nil and set the value to `""` instead

Comment (edwardmedia) [reply mmv1/third_party/terraform/utils/compute_instance_helpers.go.erb]: Added. Also added its support for `google_compute_instance_template`

File: mmv1/third_party/terraform/utils/compute_instance_helpers.go.erb [hunk 2]
```diff
@@ -248,12 +249,20 @@ func expandNetworkInterfaces(d TerraformResourceData, config *Config) ([]*comput
 			return nil, fmt.Errorf("cannot determine self_link for subnetwork %q: %s", subnetwork, err)
 		}
 
+		var nt string
+		if data["nic_type"] != nil {
```
Comment (slevenick) [inline mmv1/third_party/terraform/utils/compute_instance_helpers.go.erb:253]: Usually this is broken out into another method, similar to below `AccessConfigs: expandAccessConfigs(data["access_config"].([]interface{})),`

You could do `NicType:       expandNicType(data["nic_type"].(interface{})),` and put the nic_type specific logic in there.

**Comment by slevenick** [comment]
Is this feature GA? It looks to be beta based on the docs page: https://cloud.google.com/compute/docs/instances/create-vm-with-gvnic

We will probably need to put this behind version tags, similar to `min_node_cpus`

**Review by slevenick — COMMENTED** [review]
Looks good!

Only note would be to mark the field as Beta only in the documentation

**Review by slevenick — APPROVED** [review]

**Comment by rileykarson** [comment]
Quick note- I beleive the feature is now GA, as the addition of the instance field was an API change going from beta -> GA. Edward, if you have the time can you see if it works when we un-beta-guard all the gvnic stuff?

**Comment by ericyz** [comment]
Hi @edwardmedia , I might need to pick up your brain on the downstream module used by terraform-validator. It seems there is a conversion error caused by https://github.com/GoogleCloudPlatform/magic-modules/blob/06b0fd539a48b8fd4040321e7b0db2ecfb5cf970/mmv1/third_party/terraform/utils/compute_instance_helpers.go.erb#L258

Do you have any suggestion on how to fix it? Thanks.

CB Build Log: https://console.cloud.google.com/cloud-build/builds/1856bf87-0ce6-493b-b612-56fd3b1ef1f7?project=2843445864

**Comment by rileykarson** [comment]
fyi @melinath

**Comment by melinath** [comment]
Thanks Riley - I left a comment on https://github.com/GoogleCloudPlatform/terraform-validator/pull/195 regarding this. But it looks like this is a beta-only field :-/ that's interesting! I'll take a look.

**Comment by melinath** [comment]
Ah yeah it was moved to GA in https://github.com/GoogleCloudPlatform/magic-modules/pull/4516 so my comment on the tfv PR is correct :-)

**Pull request merged**